jax/custom_derivatives.py

```diff
@@ -951,3 +955,162 @@ def merge(l1, l2):
+
+def linear_call(fun: Callable, fun_transpose: Callable, residual_args,
+                linear_args):
+  """Call a function, indicating a custom implementation for its transpose.
+
+  The type signatures of ``fun`` and ``fun_transpose`` are:
+
+  .. code-block:: haskell
+
+    fun           :: r -> a -o b
+    fun_transpose :: r -> b -o a
+
+  where the ``-o`` arrow indicates a linear function, ``r`` is the
+  residual input type and ``a`` is the linear input type.
+
+  The functions ``fun`` and ``fun_transpose`` are coupled as
+  transposes of one another across transposition. Specifically, the
+  transpose of a ``linear_call`` primitive is another ``linear_call``
+  to ``fun_transpose``, with ``fun`` as its custom transposition.
+
+  For example, if::
+
+    def f(r, x):
+      return x / r
+
+    def t(r, t):
+      return t / r
+
+    def div_add(denom, x):
+      return x + linear_call(f, t, denom, x)
+
+    def transpose(f, x_example):
+      def transposed(y):
+        x, = jax.linear_transpose(f, x_example)(y)
+        return x
+      return transposed
+
+  Then:
+
+  >>> div_add(9., 3.)
+  12.0
+  >>> transpose(partial(div_add, 3.), 1.)(18.)  # custom
+  24.0
+  >>> transpose(lambda x: x + x / 3., 1.)(18.)  # reference
+  24.0
+
+  As another example, if::
+
+    def custom_id(x):
+      def f(_, x): return x
+      def t(_, t): return 7.
+      return linear_call(f, t, (), x)
+
+  Then:
+
+  >>> custom_id(1.)
+  1.
+  >>> transpose(custom_id, 1.)(1.)
+  7.
+  >>> transpose(transpose(custom_id, 1.), 1.)(1.)
+  1.
+  >>> transpose(transpose(transpose(custom_id, 1.), 1.), 1.)(1.)
+  7.
+
+  Args:
+    fun: a Python callable specifying a structurally linear
+      function. It should take two arguments: one of "residual" inputs
+      (type ``r``), i.e. inputs in which the function is not
+      necessarly linear, and one of "linear" inputs (type ``a``).  It
+      should return output whose components are linear in the linear
+      input (type ``b``).
+    fun_transpose: a Python callable specifying a structurally linear
+      function that is the transpose of ``fun`` with respect to its
+      linear inputs. Its first argument is the same residual inputs
+      (``r``) as ``fun``. Its second argument is of type
+      ``b``. Finally, its output is of type ``a`` and each of its
+      component are linear in its second argument (the ``b`` inputs).
+
+  Returns:
+    The call result, i.e. ``fun(residual_args, linear_args)``.
+
+  """
+  operands_res, res_tree = tree_flatten(residual_args)
+  operands_lin, lin_tree = tree_flatten(linear_args)
+
+  f_in_tree = treedef_tuple((res_tree, lin_tree))
+  f, out_tree = flatten_fun_nokwargs(lu.wrap_init(fun), f_in_tree)
+
+  res_avals = map(abstractify, operands_res)
+  lin_avals = map(abstractify, operands_lin)
+  f_jaxpr, f_consts = _initial_style_jaxpr(f, (*res_avals, *lin_avals))
+  f_jaxpr = _close_jaxpr(f_jaxpr)
+  out_avals = map(core.raise_to_shaped, f_jaxpr.out_avals)
```

Comment:
I think you and I have a TODO somewhere in the codebase to ensure these are always at shaped level, as "jaxpr types" = "abstract values at shaped level". That's not actionable for this PR; I was just reminded of it!

Reply:
Correct. It's in [partial eval of cond](https://github.com/google/jax/blob/b21341d0adb26fe2f384703a327dde8379396c64/jax/_src/lax/control_flow.py#L877).